source/retryable-reads/retryable-reads.rst

```diff
@@ -365,8 +366,11 @@ and reflects the flow described above.
       connection = server.getConnection()
     } catch (PoolClearedException poolClearedError) {
       return executeRetry(command, session, poolClearedError);
+    } catch (NetworkException networkError) {
+      // ConnectionPool should have handled SDAM state changes
```

Comment:
What does this comment mean? The block comment about `try` already talks about SDAM being updated by whatever causes `PoolClearedException` to be thrown, but it's not clear to me how a NetworkException here relates to a connection pool.

Reply:
This could use further clarity, but my intention was to differentiate poolClearedExceptions which can be caused by something else, such as a pool being in the paused state. Maybe it would be better to represent it as a flag upon the cleared exception? 

The comment mentioning SDAM is a kind of reminder that at this point the pool should've done its logic to mark the relevant server unknown. When I first put up the PR there was some changes I made to CMAP pseudo code so this related to that, we can remove it to keep the code brief.

Reply:
The diff is now entirely different, which I assume is the result of rebasing/merging atop the CSOT merge. Is `executeRetry` still the correct function to call here? I don't see it anywhere else in the code example, so I'm curious if CSOT renamed `executeRetry` to `executeRetryableRead`.

If you're confident that this line is correct feel free to just resolve the comment. I don't think I'm going to be able to wrap my head around this at the moment.

Reply:
Ok now the diff is back to what it was before the retry change. I think the tricky part about changing the pseudo code is that it's not incorrect to leave exactly how it's written now. Prior to now, drivers may have taken a poolClearedError caused by a network error during handshake and retried it because its not explcitly disallowed. I'm trying to call it out in code here as a new case along with the prose change, so even adding a flag might be confusing too.

Maybe we leave the code as is, but there could be a comment in `executeRetry` that mentions originalError might be a network error. The prose already describe this as a new case that is valid to retry, so drivers have to double check the logic that handles their pool cleared scenarios doesn't avoid retrying pool clearing caused by network issues.

Reply:
@ShaneHarvey: Just leaving this open for you to acknowledge (since it pertains more to CMAP). Feel free to resolve this thread once you've read it if you have no feedback.

Reply:
I don't see any changes to the pseudocode at all in this PR anymore. Is that expected?

Reply:
Yep, thanks for the ping on this Jeremy, so related to my comment above, I can bring back the explicit catch block but it can be confusing, a handshake failure is a "poolClearedError" so in pseudo code how best do we want to capture this new scenario? With the prose changes we've defined handshake failures as retryable and the previous form of the spec did not explicitly disallow retrying these types of failures so the code remains "correct".

If you've been taking a bit of a stab at doing this in python how does it literally look in code, do you have a flag on your error or is there some hierarchy of instanceof that captures the new case?

Reply:
A handshake failure != "poolClearedError". From my understanding poolClearedError is a very specific exception type that's only thrown when a checkout is attempted from an already paused pool. A handshake failure on the other hand is any network/command error that occurs when establishing a new connection. In fact I'd argue that a poolClearedError is one type of handshake failure.

In Python our retry logic considers and retries connection checkout errors as if the error occurred while actually running the command, eg:

```python
try:
    with pool.checkout() as connection:  # This line may raise a retryable handshake error
        executeOperation(connection)  # This line may also raise a retryable error
except Exception as exc:
    if retryable(exc):
        retry()
```

Reply:
You can take a look at what I pushed but just for ease I'll paste here:
```java
    try {
      connection = server.getConnection()
    } catch (PoolClearedException poolClearedError) {
      ...
    } catch (NetworkError networkError) {
      /* NetworkError indicates the getConnection was not able to establish
       * a successfully handshake-ed connection with the server, this is retryable.
       * We do not need to update SDAM, since network errors already mark
       * servers Unknown and empty the pool. */
      return executeRetry(command, session, poolClearedError);
    }
```
Hopefully the code comment addresses the original concern of the thread around lack of clarity, if, correct me if I'm wrong, what I've written here is accurate